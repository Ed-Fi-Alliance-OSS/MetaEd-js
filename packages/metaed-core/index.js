// @flow

// Test helper -- *** where should this go? ***
export { MetaEdTextBuilder } from './test/MetaEdTextBuilder';


// Functions

export { deepFreeze, deepFreezeAssign, prependIndefiniteArticle } from './src/Utility';
export { startingFromFileLoadP } from './src/task/Pipeline';
export { createMetaEdFile } from './src/task/MetaEdFile';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './src/task/BufferFileLoader';
export { loadFiles } from './src/task/FileSystemFilenameLoader';
export { newState } from './src/State';
export { addEntity, getAllEntities, getAllTopLevelEntities, getEntity, getTopLevelCoreEntity, getEntitiesOfType, getAllEntitiesNoSimpleTypes } from './src/model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './src/model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './src/model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './src/model/ModelType';

// Builders

export { AssociationBuilder } from './src/builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './src/builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './src/builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './src/builder/ChoiceBuilder';
export { CommonBuilder } from './src/builder/CommonBuilder';
export { CommonExtensionBuilder } from './src/builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './src/builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './src/builder/DescriptorBuilder';
export { DomainBuilder } from './src/builder/DomainBuilder';
export { DomainEntityBuilder } from './src/builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './src/builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './src/builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './src/builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './src/builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './src/builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './src/builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './src/builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './src/builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './src/builder/SharedStringBuilder';
export { StringTypeBuilder } from './src/builder/StringTypeBuilder';

// Flow types

export type { State } from './src/State';
export type { InputDirectory } from './src/task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './src/task/MetaEdFile';
export type { FileIndex } from './src/task/FileIndex';
export type { Validator } from './src/validator/Validator';
export type { ValidationFailure } from './src/validator/ValidationFailure';
export type { Enhancer } from './src/enhancer/Enhancer';
export type { EnhancerResult } from './src/enhancer/EnhancerResult';
export type { Generator } from './src/generator/Generator';
export type { GeneratorResult } from './src/generator/GeneratorResult';
export type { GeneratedOutput } from './src/generator/GeneratedOutput';
export type { MetaEdEnvironment, PluginEnvironment, SemVer } from './src/MetaEdEnvironment';
export type { MetaEdPlugin, PluginManifest } from './src/plugin/PluginTypes';

// Properties
export type { AssociationProperty } from './src/model/property/AssociationProperty';
export type { BooleanProperty } from './src/model/property/BooleanProperty';
export type { ChoiceProperty } from './src/model/property/ChoiceProperty';
export type { CommonProperty } from './src/model/property/CommonProperty';
export type { CurrencyProperty } from './src/model/property/CurrencyProperty';
export type { DateProperty } from './src/model/property/DateProperty';
export type { DecimalProperty } from './src/model/property/DecimalProperty';
export type { DescriptorProperty } from './src/model/property/DescriptorProperty';
export type { DomainEntityProperty } from './src/model/property/DomainEntityProperty';
export type { DurationProperty } from './src/model/property/DurationProperty';
export type { EntityProperty, HasReferencedEntity } from './src/model/property/EntityProperty';
export type { EnumerationProperty } from './src/model/property/EnumerationProperty';
export type { InlineCommonProperty } from './src/model/property/InlineCommonProperty';
export type { IntegerProperty } from './src/model/property/IntegerProperty';
export type { MergedProperty } from './src/model/property/MergedProperty';
export type { PercentProperty } from './src/model/property/PercentProperty';
export type { PropertyIndex } from './src/model/property/PropertyRepository';
export type { PropertyType } from './src/model/property/PropertyType';
export type { ReferentialProperty } from './src/model/property/ReferentialProperty';
export type { SchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './src/model/property/SharedShortProperty';
export type { SharedStringProperty } from './src/model/property/SharedStringProperty';
export type { ShortProperty } from './src/model/property/ShortProperty';
export type { SimpleProperty } from './src/model/property/SimpleProperty';
export type { SharedSimple } from './src/model/SharedSimple';
export type { StringProperty } from './src/model/property/StringProperty';
export type { TimeProperty } from './src/model/property/TimeProperty';
export type { YearProperty } from './src/model/property/YearProperty';

// Entities
export type { AbstractEntity } from './src/model/AbstractEntity';
export type { Association } from './src/model/Association';
export type { AssociationExtension } from './src/model/AssociationExtension';
export type { AssociationSubclass } from './src/model/AssociationSubclass';
export type { Choice } from './src/model/Choice';
export type { Common } from './src/model/Common';
export type { CommonExtension } from './src/model/CommonExtension';
export type { DecimalType } from './src/model/DecimalType';
export type { Descriptor } from './src/model/Descriptor';
export type { Domain } from './src/model/Domain';
export type { DomainEntity } from './src/model/DomainEntity';
export type { DomainEntityExtension } from './src/model/DomainEntityExtension';
export type { DomainEntitySubclass } from './src/model/DomainEntitySubclass';
export type { DomainItem } from './src/model/DomainItem';
export type { EntityRepository } from './src/model/EntityRepository';
export type { Enumeration } from './src/model/Enumeration';
export type { EnumerationItem } from './src/model/EnumerationItem';
export type { IntegerType } from './src/model/IntegerType';
export type { Interchange } from './src/model/Interchange';
export type { InterchangeExtension } from './src/model/InterchangeExtension';
export type { InterchangeItem } from './src/model/InterchangeItem';
export type { MapTypeEnumeration } from './src/model/MapTypeEnumeration';
export type { ModelBase } from './src/model/ModelBase';
export type { ModelType } from './src/model/ModelType';
export type { NamespaceInfo } from './src/model/NamespaceInfo';
export type { SchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export type { SharedDecimal } from './src/model/SharedDecimal';
export type { SharedInteger } from './src/model/SharedInteger';
export type { SharedString } from './src/model/SharedString';
export type { SourceMap } from './src/model/SourceMap';
export type { StringType } from './src/model/StringType';
export type { Subdomain } from './src/model/Subdomain';
export type { TopLevelEntity } from './src/model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { NoTopLevelEntity } from './src/model/TopLevelEntity';

// Factories
export { newMetaEdEnvironment, newPluginEnvironment } from './src/MetaEdEnvironment';
export { newMetaEdPlugin } from './src/plugin/PluginTypes';

export { newAssociationProperty } from './src/model/property/AssociationProperty';
export { newBooleanProperty } from './src/model/property/BooleanProperty';
export { newChoiceProperty } from './src/model/property/ChoiceProperty';
export { newCommonProperty } from './src/model/property/CommonProperty';
export { newCurrencyProperty } from './src/model/property/CurrencyProperty';
export { newDateProperty } from './src/model/property/DateProperty';
export { newDecimalProperty } from './src/model/property/DecimalProperty';
export { newDescriptorProperty } from './src/model/property/DescriptorProperty';
export { newDomainEntityProperty } from './src/model/property/DomainEntityProperty';
export { newDurationProperty } from './src/model/property/DurationProperty';
export { newEnumerationProperty } from './src/model/property/EnumerationProperty';
export { newInlineCommonProperty } from './src/model/property/InlineCommonProperty';
export { newIntegerProperty } from './src/model/property/IntegerProperty';
export { newMergedProperty } from './src/model/property/MergedProperty';
export { newPercentProperty } from './src/model/property/PercentProperty';
export { newSchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export { newSharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export { newSharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export { newSharedShortProperty } from './src/model/property/SharedShortProperty';
export { newSharedStringProperty } from './src/model/property/SharedStringProperty';
export { newShortProperty } from './src/model/property/ShortProperty';
export { newSimpleProperty } from './src/model/property/SimpleProperty';
export { newStringProperty } from './src/model/property/StringProperty';
export { newTimeProperty } from './src/model/property/TimeProperty';
export { newYearProperty } from './src/model/property/YearProperty';

export { newAbstractEntity } from './src/model/AbstractEntity';
export { newAssociation } from './src/model/Association';
export { newAssociationExtension } from './src/model/AssociationExtension';
export { newAssociationSubclass } from './src/model/AssociationSubclass';
export { newChoice } from './src/model/Choice';
export { newCommon } from './src/model/Common';
export { newCommonExtension } from './src/model/CommonExtension';
export { newDecimalType } from './src/model/DecimalType';
export { newDescriptor } from './src/model/Descriptor';
export { newDomain } from './src/model/Domain';
export { newDomainEntity } from './src/model/DomainEntity';
export { newDomainEntityExtension } from './src/model/DomainEntityExtension';
export { newDomainEntitySubclass } from './src/model/DomainEntitySubclass';
export { newDomainItem } from './src/model/DomainItem';
export { newEnumeration } from './src/model/Enumeration';
export { newEnumerationItem } from './src/model/EnumerationItem';
export { newInlineCommon } from './src/model/Common';
export { newIntegerType } from './src/model/IntegerType';
export { newInterchange } from './src/model/Interchange';
export { newInterchangeExtension } from './src/model/InterchangeExtension';
export { newInterchangeItem } from './src/model/InterchangeItem';
export { newMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { newNamespaceInfo } from './src/model/NamespaceInfo';
export { newSchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export { newSharedDecimal } from './src/model/SharedDecimal';
export { newSharedInteger } from './src/model/SharedInteger';
export { newSharedString } from './src/model/SharedString';
export { newSourceMap } from './src/model/SourceMap';
export { newStringType } from './src/model/StringType';
export { newSubdomain } from './src/model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './src/model/TopLevelEntity';
