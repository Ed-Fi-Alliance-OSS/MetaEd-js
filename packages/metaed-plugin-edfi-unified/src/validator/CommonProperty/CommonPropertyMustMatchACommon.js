// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.common.forEach(property => {
    if (metaEd.entity.common.get(property.metaEdName) == null) {
      failures.push({
        validatorName: 'CommonPropertyMustMatchACommon',
        category: 'error',
        message: `Common property '${property.metaEdName}' does not match any declared Common.`,
        sourceMap: property.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}
