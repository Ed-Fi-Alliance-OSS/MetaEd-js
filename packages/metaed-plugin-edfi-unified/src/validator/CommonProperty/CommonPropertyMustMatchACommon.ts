import { MetaEdEnvironment, ValidationFailure, ModelBase } from 'metaed-core';
import { getEntityFromNamespaceChain } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.propertyIndex.common.forEach(property => {
    const referencedEntity: ModelBase | null = getEntityFromNamespaceChain(
      property.metaEdName,
      property.referencedNamespaceName,
      property.namespace,
      'common',
    );

    if (referencedEntity == null) {
      failures.push({
        validatorName: 'CommonPropertyMustMatchACommon',
        category: 'error',
        message: `Common property '${property.metaEdName}' does not match any declared Common in namespace ${
          property.referencedNamespaceName
        }.`,
        sourceMap: property.sourceMap.type,
        fileMap: null,
      });
    }
  });

  return failures;
}
