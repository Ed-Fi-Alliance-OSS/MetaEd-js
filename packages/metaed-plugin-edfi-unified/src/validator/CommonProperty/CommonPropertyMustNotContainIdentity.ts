// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.propertyIndex.common.forEach((common) => {
    if (!common.isPartOfIdentity) return;
    failures.push({
      validatorName: 'CommonPropertyMustNotContainIdentity',
      category: 'error',
      message: `Common property ${common.metaEdName} is invalid to be used for the identity of ${common.parentEntity.typeHumanizedName} ${common.parentEntityName}.`,
      sourceMap: common.sourceMap.isPartOfIdentity,
      fileMap: null,
    });
  });
  return failures;
}
