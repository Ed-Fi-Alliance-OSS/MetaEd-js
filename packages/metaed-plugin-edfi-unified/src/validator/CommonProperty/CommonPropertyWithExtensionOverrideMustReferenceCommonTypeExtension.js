// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { CommonPropertySourceMap } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.common.forEach(common => {
    if (!common.isExtensionOverride || (common.isExtensionOverride && metaEd.entity.commonExtension.has(common.metaEdName))) return;
    failures.push({
      validatorName: 'CommonPropertyWIthExtensionOverrideMustReferenceCommonTypeExtension',
      category: 'error',
      message: `'common extension' is invalid for property ${common.metaEdName} on ${common.parentEntity.typeHumanizedName} ${common.parentEntity.metaEdName}. 'common extension' is only valid for referencing Common extensions.`,
      sourceMap: ((common.sourceMap: any): CommonPropertySourceMap).isExtensionOverride,
      fileMap: null,
    });
  });
  return failures;
}
