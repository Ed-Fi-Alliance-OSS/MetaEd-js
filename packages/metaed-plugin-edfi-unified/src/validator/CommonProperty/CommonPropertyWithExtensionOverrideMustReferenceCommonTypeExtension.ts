import { MetaEdEnvironment, ValidationFailure, CommonPropertySourceMap, ModelBase } from '@edfi/metaed-core';
import { getEntityFromNamespaceChain } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.propertyIndex.common.forEach((property) => {
    if (!property.isExtensionOverride) return;
    const referencedEntity: ModelBase | null = getEntityFromNamespaceChain(
      property.metaEdName,
      // implicitly we reference Common Extension declared in same namespace as property using it
      property.namespace.namespaceName,
      property.namespace,
      'commonExtension',
    );

    if (referencedEntity == null) {
      failures.push({
        validatorName: 'CommonPropertyWithExtensionOverrideMustReferenceCommonTypeExtension',
        category: 'error',
        message: `No common extension named ${property.metaEdName} is declared in namespace ${property.referencedNamespaceName}.`,
        sourceMap: (property.sourceMap as CommonPropertySourceMap).isExtensionOverride,
        fileMap: null,
      });
    }
  });
  return failures;
}
