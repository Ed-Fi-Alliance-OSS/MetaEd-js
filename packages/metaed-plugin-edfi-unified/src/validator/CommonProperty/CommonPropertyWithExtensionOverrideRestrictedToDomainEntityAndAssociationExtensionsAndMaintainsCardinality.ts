import {
  CommonPropertySourceMap,
  EntityProperty,
  TopLevelEntity,
  MetaEdEnvironment,
  ModelType,
  ValidationFailure,
  Namespace,
} from '@edfi/metaed-core';
import { getEntityFromNamespaceChain } from '@edfi/metaed-core';

const validEntityTypes: ModelType[] = ['domainEntityExtension', 'associationExtension'];

function cardinalitiesMatch(originalProperty: EntityProperty, overrideProperty: EntityProperty): boolean {
  return (
    (originalProperty.isRequired && overrideProperty.isRequired) ||
    (originalProperty.isOptional && overrideProperty.isOptional) ||
    (originalProperty.isRequiredCollection && overrideProperty.isRequiredCollection) ||
    (originalProperty.isOptionalCollection && overrideProperty.isOptionalCollection)
  );
}

function baseEntityMatchingProperty(namespace: Namespace, overrideProperty: EntityProperty): EntityProperty | undefined {
  if (!validEntityTypes.includes(overrideProperty.parentEntity.type)) return undefined;
  // base type being extended - sketchy string manipulation here
  const baseType: ModelType = overrideProperty.parentEntity.type.replace('Extension', '') as ModelType;
  const baseSubclassType: ModelType = `${baseType}Subclass` as ModelType;

  const baseEntity: TopLevelEntity | null = getEntityFromNamespaceChain(
    overrideProperty.parentEntity.baseEntityName,
    overrideProperty.parentEntity.baseEntityNamespaceName,
    namespace,
    baseSubclassType,
    baseType,
  ) as TopLevelEntity | null;

  if (baseEntity == null) return undefined;

  return baseEntity.properties.find(
    (property) =>
      property.metaEdName === overrideProperty.metaEdName &&
      property.referencedNamespaceName === overrideProperty.referencedNamespaceName &&
      property.type === overrideProperty.type,
  );
}

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.propertyIndex.common.forEach((commonProperty) => {
    if (!commonProperty.isExtensionOverride) return;
    const baseMatchingProperty: EntityProperty | undefined = baseEntityMatchingProperty(
      commonProperty.namespace,
      commonProperty,
    );
    if (baseMatchingProperty == null) {
      failures.push({
        validatorName:
          'CommonPropertyWithExtensionOverrideRestrictedToDomainEntityAndAssociationExtensionsAndMaintainsCardinality',
        category: 'error',
        message: `'common extension' is invalid for property ${commonProperty.metaEdName} on ${commonProperty.parentEntity.typeHumanizedName} ${commonProperty.parentEntity.metaEdName}. 'common extension' is only valid for referencing common properties with the same declaration as in the base entity.`,
        sourceMap: (commonProperty.sourceMap as CommonPropertySourceMap).isExtensionOverride,
        fileMap: null,
      });
    } else if (!cardinalitiesMatch(baseMatchingProperty, commonProperty)) {
      failures.push({
        validatorName:
          'CommonPropertyWithExtensionOverrideRestrictedToDomainEntityAndAssociationExtensionsAndMaintainsCardinality',
        category: 'error',
        message: `'common extension' is invalid for property ${commonProperty.metaEdName} on ${commonProperty.parentEntity.typeHumanizedName} ${commonProperty.parentEntity.metaEdName}. 'common extension' must maintain the same cardinality as in the base entity.`,
        sourceMap: (commonProperty.sourceMap as CommonPropertySourceMap).isExtensionOverride,
        fileMap: null,
      });
    }
  });
  return failures;
}
