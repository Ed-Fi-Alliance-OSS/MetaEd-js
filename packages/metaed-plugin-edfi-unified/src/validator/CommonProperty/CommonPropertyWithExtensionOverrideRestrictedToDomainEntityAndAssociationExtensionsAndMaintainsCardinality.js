// @flow
import type {
  EntityProperty,
  EntityRepository,
  MetaEdEnvironment,
  ModelType,
  ValidationFailure } from '../../../../metaed-core/index';
import { CommonPropertySourceMap } from '../../../../metaed-core/src/model/property/CommonProperty';

const validEntityTypes: ModelType[] = [
  'domainEntityExtension',
  'associationExtension',
];

function cardinalitiesMatch(originalProperty: EntityProperty, overrideProperty: EntityProperty): boolean {
  return ((originalProperty.isRequired && overrideProperty.isRequired)
    || (originalProperty.isOptional && overrideProperty.isOptional)
    || (originalProperty.isRequiredCollection && overrideProperty.isRequiredCollection)
    || (originalProperty.isOptionalCollection && overrideProperty.isOptionalCollection));
}

function parentEntityProperty(entity: EntityRepository, overrideProperty: EntityProperty): EntityProperty | void {
  if (!validEntityTypes.includes(overrideProperty.parentEntity.type)) return undefined;
  const parentType = overrideProperty.parentEntity.type.replace('Extension', '');
  // $FlowIgnore - allowing parentType to specify the entityRepository Map property
  const parentEntity = entity[parentType].get(overrideProperty.parentEntityName);
  return parentEntity.properties.find(property =>
      property.metaEdName === overrideProperty.metaEdName
      && property.type === overrideProperty.type);
}

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.common.forEach(common => {
    if (!common.isExtensionOverride) return;
    const parentProperty = parentEntityProperty(metaEd.entity, common);
    if (parentProperty && cardinalitiesMatch(parentProperty, common)) return;
    failures.push({
      validatorName: 'CommonPropertyWithExtensionOverrideRestrictedToDomainEntityAndAssociationExtensionsAndMaintainsCardinality',
      category: 'error',
      message: `'common extension' is invalid for property ${common.metaEdName} on ${common.parentEntity.typeHumanizedName} ${common.parentEntity.metaEdName}. 'common extension' is only valid for referencing Common extensions.`,
      sourceMap: ((common.sourceMap: any): CommonPropertySourceMap).isExtensionOverride,
      fileMap: null,
    });
  });
  return failures;
}
