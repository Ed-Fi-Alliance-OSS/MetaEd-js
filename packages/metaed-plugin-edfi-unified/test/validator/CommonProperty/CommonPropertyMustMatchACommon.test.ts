import { newMetaEdEnvironment, MetaEdTextBuilder, CommonBuilder, DomainEntityBuilder, NamespaceBuilder } from 'metaed-core';
import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/CommonProperty/CommonPropertyMustMatchACommon';

describe('when common property has identifier of common', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'DomainEntityName';
  const entityName = 'EntityName';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartCommon(entityName)
      .withDocumentation('doc')
      .withStringProperty('StringProperty', 'doc', true, false, '100')
      .withEndCommon()

      .withStartDomainEntity(domainEntityName)
      .withDocumentation('doc')
      .withCommonProperty(entityName, 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should have no validation failures()', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when common property has invalid identifier', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'DomainEntityName';
  const entityName = 'EntityName';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartCommon('WrongName')
      .withDocumentation('doc')
      .withStringProperty('StringProperty', 'doc', true, false, '100')
      .withEndCommon()

      .withStartDomainEntity(domainEntityName)
      .withDocumentation('doc')
      .withCommonProperty(entityName, 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    failures = validate(metaEd);
  });

  it('should have validation failures()', () => {
    expect(failures).toHaveLength(1);
  });

  it('should have validation failure for property', () => {
    expect(failures[0].validatorName).toBe('CommonPropertyMustMatchACommon');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot(
      'when common property has invalid identifier should have validation failures for each property -> message ',
    );
    expect(failures[0].sourceMap).toMatchSnapshot(
      'when common property has invalid identifier should have validation failures for each property -> sourceMap',
    );
  });
});

describe('when common property has identifier of common in dependency namespace', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'DomainEntityName';
  const entityName = 'EntityName';
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;
  let extensionNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartCommon(entityName)
      .withDocumentation('doc')
      .withStringProperty('StringProperty', 'doc', true, false, '100')
      .withEndCommon()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'ProjectExtension')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('doc')
      .withCommonProperty(`EdFi.${entityName}`, 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespace = metaEd.namespace.get('Extension');
    extensionNamespace.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should have no validation failures()', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when common property has invalid identifier of common in dependency namespace', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'DomainEntityName';
  const entityName = 'EntityName';
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;
  let extensionNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartCommon(entityName)
      .withDocumentation('doc')
      .withStringProperty('StringProperty', 'doc', true, false, '100')
      .withEndCommon()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'ProjectExtension')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('doc')
      .withCommonProperty('NotValid', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespace = metaEd.namespace.get('Extension');
    extensionNamespace.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should have validation failures()', () => {
    expect(failures).toHaveLength(1);
  });

  it('should have validation failure for property', () => {
    expect(failures[0].validatorName).toBe('CommonPropertyMustMatchACommon');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});

// can't reference entities outside of dependency list
describe('when common property refers to common in non-dependency namespace', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;
  let extensionNamespacea: any = null;
  let extensionNamespaceb: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartDomainEntity('CoreEntity')
      .withDocumentation('doc')
      .withBooleanProperty('Dummy', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extensiona', 'ProjectExtensiona')
      .withStartDomainEntity('ExtensionEntity')
      .withDocumentation('doc')
      .withCommonProperty('CommonEntity', 'doc', true, false)
      .withEndDomainEntity()
      .withEndNamespace()

      .withBeginNamespace('Extensionb', 'ProjectExtensionb')
      .withStartCommon('CommonEntity')
      .withDocumentation('doc')
      .withStringProperty('StringProperty', 'doc', true, false, '100')
      .withEndCommon()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new CommonBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespacea = metaEd.namespace.get('Extensiona');
    extensionNamespaceb = metaEd.namespace.get('Extensionb');
    extensionNamespacea.dependencies.push(coreNamespace);
    extensionNamespaceb.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should have validation failures()', () => {
    expect(failures).toHaveLength(1);
  });

  it('should have validation failure for property', () => {
    expect(failures[0].validatorName).toBe('CommonPropertyMustMatchACommon');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});
