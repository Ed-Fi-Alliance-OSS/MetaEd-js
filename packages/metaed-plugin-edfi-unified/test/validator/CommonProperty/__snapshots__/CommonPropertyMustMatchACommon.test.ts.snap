// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when common property has invalid identifier of common in dependency namespace should have validation failure for property 1`] = `"Common property 'NotValid' does not match any declared Common in namespace Extension."`;

exports[`when common property has invalid identifier of common in dependency namespace should have validation failure for property 2`] = `
Object {
  "column": 4,
  "line": 15,
  "tokenText": "common",
}
`;

exports[`when common property has invalid identifier should have validation failure for property: when common property has invalid identifier should have validation failures for each property -> message  1`] = `"Common property 'EntityName' does not match any declared Common in namespace EdFi."`;

exports[`when common property has invalid identifier should have validation failure for property: when common property has invalid identifier should have validation failures for each property -> sourceMap 1`] = `
Object {
  "column": 4,
  "line": 13,
  "tokenText": "common",
}
`;

exports[`when common property refers to common in non-dependency namespace should have validation failure for property 1`] = `"Common property 'CommonEntity' does not match any declared Common in namespace Extensiona."`;

exports[`when common property refers to common in non-dependency namespace should have validation failure for property 2`] = `
Object {
  "column": 4,
  "line": 14,
  "tokenText": "common",
}
`;
